### PR TITLE
Feature/us7507 cms from maestro

### PR DESCRIPTION
**Pull request by richardbarker**

### CODE REVIEW

File: apps/crossroads_interface/web/plugs/meta_plug.ex [hunk 1]
```diff
@@ -5,43 +5,88 @@ defmodule CrossroadsInterface.Plug.Meta do
...
+
   def init(default), do: default
 
   def call(conn, _default) do
-    path = conn.request_path
-           |> String.split("/")
-           |> Enum.filter(&(&1 != ""))
-           |> Enum.join(".")
+    
+    # if(conn.request_path |> ContentHelpers.addTrailingSlashIfNecessary |> Pages.page_exists?) do
```
Comment (mjvcallibrity) [inline apps/crossroads_interface/web/plugs/meta_plug.ex]: Do we need to hang on to all of this commented out code?

File: apps/crossroads_content/lib/cms_client.ex [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+  end
+
+  @doc false
+  def handle_call({:all, url, params}, _from, state) do
+    path = "#{url}?#{URI.encode_query(params)}"
+    make_call(path, state)
+  end
+
+  @doc false
+  defp make_call(path, state) do
```
Comment (silbermm) [inline apps/crossroads_content/lib/cms_client.ex]: I agree this method is rather large and hard to read

Comment (silbermm) [inline apps/crossroads_content/lib/cms_client.ex]: i would do something like....
```
result = match_response(HTTPoison.get("#{@base_url}/api/#{path}",["Accept": "application/json"], [recv_timeout: @timeout]))




defp match_response({:ok, %HTTPoison.Response{status_code: 404, body: body}}) do
  {:error, 404, decode_request(Poison.decode(body))}
end
defp match_response({:ok, %HTTPoison.Response{status_code: 200, body: body, headers: headers}}) do
  if determine_headers(headers) do
     {:error, 400, %{}}
  else 
    {:ok, 200, decode_request(Poison.decode(body))}
   end
end
defp match_response({:error, %HTTPoison.Error{reason: reason}}) do
  {:error, 500, %{error: reason}}
end
defp match_response(_), do: {:error, 0, %{error: "unknown response"}}

**Comment by silbermm** [comment]
Also I think that @danrye and/or @IngageKrizJ  should understand what this is doing on startup and sign off on it. It could possibly benefit from load testing.

File: apps/crossroads_content/test/cms_client_test.exs [hunk 3]
```diff
@@ -39,7 +45,7 @@ defmodule CrossroadsContentPagesTest do
 
   test "get content blocks" do
```
Comment (richardbarker) [inline apps/crossroads_content/test/cms_client_test.exs]: There are additional tests on lines 109 and 116 to test the caching around this call.

File: apps/crossroads_interface/test/controllers/cms_page_controller_test.exs [hunk 4]
```diff
@@ -0,0 +1,78 @@
+defmodule CrossroadsInterface.CmsPageControllerTest do
+  use CrossroadsInterface.ConnCase
+  alias CrossroadsContent.CmsClient
+  alias CrossroadsContent.Pages
+  import Mock
+
+  require IEx
+
+  @get_page_response {:ok,
```
Comment (richardbarker) [inline apps/crossroads_interface/test/controllers/cms_page_controller_test.exs]: I like it here.  If I'm writing tests I can see exactly what I'm testing against.

File: apps/crossroads_interface/package.json [hunk 5]
```diff
@@ -8,7 +8,7 @@
     "repository": {},
     "dependencies": {
         "crds-shared-header": "0.2.11",
-        "crds-styles": "0.5.0",
+        "crds-styles": "crdschurch/crds-styles#development",
```
Comment (richardbarker) [inline apps/crossroads_interface/package.json]: This is only temporary.  We'll switch to a fixed version before deployment.

File: apps/crossroads_content/lib/cms_client.ex [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        Cachex.set(:cms_cache, path, result)
+      end      
+    end
+    {:reply, result, state}
+  end
+
+  defp match_response({:ok, %HTTPoison.Response{status_code: 404, body: body}}) do
+    {:error, 404, decode_request(Poison.decode(body))}
+  end
+
```
Comment (silbermm) [inline apps/crossroads_content/lib/cms_client.ex]: remove empty lines when using pattern matching in arguments see [the style guide](https://github.com/christopheradams/elixir_style_guide#single-line-defs)

**Comment by silbermm** [comment]
In general this code works. My biggest concern is that it makes a big impact on app startup. The app is unresponsive until the Cms pages are loaded which is anywhere from 20 to 30 seconds in int. And what happens when the CMS becomes unresponsive while trying to pull this data? I wish there were a way to just get all the Pages urls and store them which would be a much quicker call and smaller payload. 

I still think that @IngageKrizJ  and/or @danrye  need to give the thumbs up before merging.

**Comment by danrye** [comment]
Replying via email without reviewing, just seeing the comments.  

What about a persisted version of this is always saved, then on startup it uses that persisted version and the the refresh (which you'd have run right away) will pickup changes?  The refresh also updates the persisted version?  This also gives you a nice backup if the CMS goes down.

**Pull request merged**